### PR TITLE
Working on 1 1 3 hello d3d11 explain input element desc better #95 

### PR DESCRIPTION
**Pull request by 16-Bit-Dog**

### CODE REVIEW

**Review by deccer — COMMENTED** [review]
any reason for this empty commit?

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 1]
```diff
@@ -373,14 +374,21 @@ D3D11_INPUT_ELEMENT_DESC vertexInputLayoutInfo[] is an array since when we pass
...
+
 - `AlignedByteOffset`: the offset of where this field starts in the memory of 1 vertex in bytes. For example with `vertexInputLayoutInfo`, since we want `COLOR` after `POSITION` which is a `float3` (12 bytes) —  we tell D3D11 that we want a 12 byte offset to put `COLOR` immediately after `POSITION` in memory. `POSITION` on the other hand needs no offset and is the first element (offset is 0). 
-    offsetof(VertexPositionColor, color) is a preprocessor that calcualtes the offset for us using a struct and its element name as input
-- `InputSlotClass`: The rate of input is either per-vertex or per-instance, we don't care about instances right now, so we'll set this to PER_VERTEX. 
+    `offsetof(VertexPositionColor, color)` is a preprocessor that calcualtes the offset for us of an element using a struct and the desiered element name as input — this case we inputted the struct `VertexPositionColor` and get the offset of the element `color`. This is why we did not need to type out `12` or `0` in the `AlignedByteOffset` ourselves.
```
Comment (deccer) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md]: instead of "is a preprocessor that calcualtes the offset for us of an element using a struct and the desiered element name as input"

"is a macro that`calculates` the offset of fields within a struct via the the field name"

You could also explain here why the outcome would be 12, 24 etc, by explaining the size of he fields etc

you really need to look at typos my dude :) they are usually left with red squiggles in all major editors these days

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 2]
```diff
@@ -404,14 +388,15 @@ struct VSInput
...
-then create an input layout exactly with this format.
+The vertex shader expects per vertex: 2 vectors of 3 (4 byte) components. 
 
-First, creating a `struct` in our C++ source with that maps to our `VSInput` as 
-closely as possible will make our life easier.
+We should then create an input layout exactly with this format:
 
-To do this we'll use DirectXMath which has types that map perfectly to HLSL, both
-of our inputs are `float3` in HLSL, which means that this translates to `DirectX::XMFLOAT3`.
+First of all, creating a `struct` in our C++ source with the same memory organization as our `VSInput` will make our life easier when imagining how 1 vertex will fit on the GPU.
```
Comment (deccer) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:396]: "memory organization" is a confusing term here... maybe "field layout" would work or "order of fields"?

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 3]
```diff
@@ -404,14 +388,15 @@ struct VSInput
...
+
+To do this we'll use DirectXMath which has types that map perfectly to HLSL, both of our inputs are `float3` in HLSL, which means that this translates to `DirectX::XMFLOAT3`
```
Comment (deccer) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:398]: can you wrap DirectXMath with code backticks?

**Review by deccer — CHANGES_REQUESTED** [review]
A few things to consider/change probably.

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 4]
```diff
@@ -14,19 +21,29 @@ All the steps in the graphics pipeline go from top to bottom and are shown below
...
 
-The **Vertex Shader** is the stage where our vertices are processed however we want, although we don't do much processing here, and in the end they're transformed to [screen-space coordinates](link to the coordinate system chapter?)
+The **Vertex Shader** is the stage where our vertices are processed however 
+we want, although we don't do much processing here, and in the end they're 
+transformed to [screen-space coordinates](link to the coordinate system chapter?)
 
-The vertices are usually read from a **Vertex Buffer** which are laid out in a particular way. The vertex shader will be run however many times we tell it to run, which is specified in the first parameter of `ID3D11DeviceContext::Draw()` (more on this later), for instance if we call `Draw(3, 0)`, the vertex shader will run 3 times.
+The vertices are usually read from a **Vertex Buffer** which are laid out in 
+a particular way. The vertex shader will be run however many times we tell
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:46]: Maybe mention the possibilities of not doing it like this without saying exactly how (because stuff like Vertex Pulling exists);
> The vertices are usually read from a **Vertex Buffer** which are laid out in 
a particular way.

->

"Though not required specifically, the easiest method for retrieving vertices is from a **Vertex Buffer**, this is a buffer containing all vertices laid out in a specific way. There are other more advanced methods but we'll leave those out for the sake of simplicity for now."

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 5]
```diff
@@ -14,19 +21,29 @@ All the steps in the graphics pipeline go from top to bottom and are shown below
...
 
-The vertices are usually read from a **Vertex Buffer** which are laid out in a particular way. The vertex shader will be run however many times we tell it to run, which is specified in the first parameter of `ID3D11DeviceContext::Draw()` (more on this later), for instance if we call `Draw(3, 0)`, the vertex shader will run 3 times.
+The vertices are usually read from a **Vertex Buffer** which are laid out in 
+a particular way. The vertex shader will be run however many times we tell 
+it to run, which is specified in the first parameter of `ID3D11DeviceContext::Draw()` 
+(more on this later), for instance if we call `Draw(3, 0)`, the vertex shader will 
+run 3 times.
 
-Since we only want to draw a triangle, we don't need to do much processing, we can just provide the input vertices as the output.
+Since we only want to draw a triangle, we don't need to do much processing,
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:51]: > Since we only want to draw a triangle, we don't need to do much processing,

overly simplified in my opinion, this gives the reader the notion that each triangle only needs direct forwarding to the PS, it would be more correct to note that "Since the vertex positions we supply are already in NDC space we don't need to transform/process them for the rasterisation stage, so we can just provide the input vertices as the output"

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 6]
```diff
@@ -14,19 +21,29 @@ All the steps in the graphics pipeline go from top to bottom and are shown below
...
+run 3 times.
 
-Since we only want to draw a triangle, we don't need to do much processing, we can just provide the input vertices as the output.
+Since we only want to draw a triangle, we don't need to do much processing, 
+we can just provide the input vertices as the output.
 
-The vertex buffer can be omitted, for example if we want to draw a full screen triangle, instead of creating a vertex buffer, for simplicity we can just hardcode the vertices in the vertex shader without having to bind a vertex buffer.
+The vertex buffer can be omitted, for example if we want to draw a full screen 
+triangle, instead of creating a vertex buffer, for simplicity we can just hardcode 
+the vertices in the vertex shader without having to bind a vertex buffer.
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:56]: slightly too advanced to mention here in my opinion, this requires drawing without a vertex buffer and only relying on SV_VertexID as well as an algorithm to generate a unique vertex per ID, I think this is something we should mention once we start more advanced techniques such as vertex pulling where this comes into play a bit more, perhaps leave this all out?

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 7]
```diff
@@ -54,27 +71,43 @@ VSOutput Main(VSInput input)
 }
 ```
 
-First off, we define 2 types, `VSInput` and `VSOutput` which represent the vertex shader's input and output.
+First off, we define 2 types, `VSInput` and `VSOutput` which represent the 
+vertex shader's input and output.
+
+The input is 2, `float3` (vector of 3 float components), the first is the 
+"position" field, which are coordinates ranging from [-1.0, 1.0], if the values 
+are outside this range they are clipped, and we won't see them on screen.
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:89]: This is entirely false, the input vertices have no requirement on position or any data at all really,  this clipping stage only happens on vertices from the SV_Position (output) after the vertex shader is ran, also when noting [-1.0,1.0] ranges please mention NDC / Normalized Device Coordinates and/or explain it

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 8]
```diff
@@ -54,27 +71,43 @@ VSOutput Main(VSInput input)
...
-The input is 2, `float3` (vector of 3 float components), the first is the "position" field, which are coordinates ranging from [-1.0, 1.0], if the values are outside this range they are clipped, and we won't see them on screen.
+The second is the "color" field, which we also pass as the output of this stage 
+onto the pixel shader.
 
-The second is the "color" field, which we also pass as the output of this stage onto the pixel shader.
+Notice how all our fields have a colon and some identifier attached to them, 
+these are "semantics". Semantics that are preceded by `SV` are called 
+"system-value semantics" and their meaning and usage is defined by D3D11. 
+`SV_Position` for example means that the field `position` will be used by 
+D3D11 as the actual output of the vertex shader.
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:98]: Mention NDC after SV_Position and what it does for the entire graphics pipeline

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 9]
```diff
@@ -54,27 +71,43 @@ VSOutput Main(VSInput input)
...
 
-Notice how all our fields have a colon and some identifier attached to them, these are "semantics".
-Semantics that are preceded by `SV` are called "system-value semantics" and their meaning and usage is defined by D3D11. `SV_Position` for example means that the field `position` will be used by D3D11 as the actual output of the vertex shader.
-Everything else are "user defined semantics" and their naming is up to us. These are used to pass data between shader stages.
+Everything else are "user defined semantics" and their naming is up to us. 
+These are used to pass data between shader stages.
 
-Then we have our `VSOutput`, which has our vertices in the first field `position` and our color in the second field `color`.
+Then we have our `VSOutput`, which has our vertices in the first field `position`
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:103]: here we should mention SV_Position, NDC and potentially its effect on the zbuffer

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 10]
```diff
@@ -54,27 +71,43 @@ VSOutput Main(VSInput input)
...
 
-Then we have our `VSOutput`, which has our vertices in the first field `position` and our color in the second field `color`.
+Then we have our `VSOutput`, which has our vertices in the first field `position` 
+and our color in the second field `color`.
 
-Finally, we have our main function, which takes in a single parameter which is our input in the form of `VSInput`, and returns our output in the form of a `VSOutput`. Since we don't do any processing, we simply make a new instance of `VSOutput`, initialize it all to 0 and forward our input position and color to the output.
+Finally, we have our main function, which takes in a single parameter which is 
+our input in the form of `VSInput`, and returns our output in the form of a `VSOutput`.
+ Since we don't do any processing, we simply make a new instance of `VSOutput`, 
+ initialize it all to 0 and forward our input position and color to the output.
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:109]: splitting hairs, but having only structures in/output isn't required, you can mention each input seperately and directly declare output as well:
eg;
`float4 main(float3 inPos : POSITION, float2 inUV: TEXCOORD) : SV_Position`

Not sure we want to mention this, just adding information here

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 11]
```diff
@@ -54,27 +71,43 @@ VSOutput Main(VSInput input)
...
 
 ### Pixel Shader
 
-The **Pixel Shader** is the stage where we give the pixels on our render target color, it is invoked for each pixel that is covered by a triangle.
+The **Pixel Shader** is the stage where we give the pixels on our render target 
+color, it is invoked for each pixel that is covered by a triangle.
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:114]: being nitpicky on my side here, but this isn't 100% accurate since the GPU works in at least 2x2 tiles, so even if a pixel isn't covered by the rasterized triangle, it could still run a PS for it.., just covering bases here, might be worth mentioning, though I guess also rather advanced knowledge?

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 12]
```diff
@@ -102,19 +135,35 @@ PSOutput Main(PSInput input)
 
 Here as well we have an input `PSInput`, and an output `PSOutput`.
 
-Since there aren't any other shaders in between the VS and the PS, the VS's output is the PS's input, the naming might be a bit confusing but that's the gist of it, PSInput should match the VSOutput in vertex shader, this isn't entirely required but not doing so is only advisable if you really know what you are doing.
+Since there aren't any other shaders in between the VS and the PS, the VS's
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:148]: > Since there aren't any other shaders in between the VS and the PS

Perhaps reword this to: "Since we haven't configured any other shader stages in between the VS and PS ..."
as this implies there are no other stages, even though there are quite a few left :)

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 13]
```diff
@@ -102,19 +135,35 @@ PSOutput Main(PSInput input)
...
-`D3D11` lets us write up to 8 render targets simultaneously from the same pixel shader, those come in handy when implementing more advanced shading techniques, for example a popular technique that uses 4 or more
+Notice how we have another semantic string attached to the `color` field, 
+this semantic string specifies which render target we want to be writing to, 
+the `0` after `SV_Target` is the index of our render target, in our case, 
+we have only one, so we write `SV_Target0` or `SV_Target`.
 
-And lastly, our `Main` function, following the same pattern as in the VS, we have one parameter, the input, and one return value, the output, again we create an instance of `PSOutput`, initialize everything to 0, and write the color we got from the input, to our output.
+`D3D11` lets us write up to 8 render targets simultaneously from the same pixel 
+shader, those come in handy when implementing more advanced shading techniques, 
+for example a popular technique that uses 4 or more
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:166]: > for example a popular technique that uses 4 or more

too vague, better to remove this entire sentence as we're not diving into deferred rendering (which I assume this is referring to?), which also doesn't specifically require 4 or more

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 14]
```diff
@@ -390,9 +505,13 @@ Now all we have to do is create a vertex buffer (don't worry it's really easy) a
...
+buffer that resides in our device's memory, which means really fast access. This 
+buffer will be then bound and read by the vertex shader.
 
-Creating a vertex buffer is also really easy, first we have to make some data to put in our buffer, since we want to draw a triangle, we will be creating 3 vertices using our `VertexPositionColor` struct.
+Creating a vertex buffer is also really easy, first we have to make some data to put 
+in our buffer, since we want to draw a triangle, we will be creating 3 vertices using 
+our `VertexPositionColor` struct.
 
 ```cpp
 constexpr VertexPositionColor vertices[] =
```
Comment (Themperror) [inline docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:546]: use std::array instead? also gets rid of this atrocious `_countof` which I don't believe is portable at all?  (because we can use `.size()` then)

**Review by Themperror — CHANGES_REQUESTED** [review]

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 10 above)
Comment (deccer) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:109]: Yep this is possible, should definitly be mentioned, but it seems kinda convention that you use structs - its more readable IMHO (mileage my wary)

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 14 above)
Comment (deccer) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:546]: If this is changed in the docs, it needs to be changed in code as well.

Mhh not sure how much the comments divert from the actual scope of this PR. This should be addressed in a different issue/PR

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 15]
```diff
@@ -1,11 +1,18 @@
...
+the Device and the SwapChain, but simply clearing the Window with 
+some color is pretty boring.
+
 This time we'll be drawing our first triangle with a nice froge-like color.
 
 ## The Pipeline
 
-The fundamental part of all graphic APIs is the "Graphics Pipeline". Everything from a single triangle, textured frog or the whole Elden Ring map goes through this pipeline. It is a series of functions that either exist in hardware, can be configured or are fully programmable. It transforms everything we draw in 3D space to the 2D space that is our monitor.
+The fundamental part of all graphic APIs is the "Graphics Pipeline". Everything 
+from a single triangle, textured frog or the whole Elden Ring map goes through
```
[the remark this answers is not in the document]
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:12]: "`Almost everything` from a single triangle, ..." is more correct, I do agree

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md [hunk 16]
```diff
@@ -14,19 +21,29 @@ All the steps in the graphics pipeline go from top to bottom and are shown below
...
 
-The Vertex and Pixel shaders are fully programmable, and we'll write a very basic program for them.
+The Vertex and Pixel shaders are fully programmable and we'll write a very basic program for them.
 
-The other two stages are not programmable, but they are fairly easy to understand and configure, the **Input Assembler** is responsible for processing the vertices in an eventual vertex buffer into the primitive of our choice, which is of course, triangles in our case, and sending this processed output to the Vertex Shader. The **Output Merger** instead is responsible for combining the values written by the pixel shader, may that be depth, color or other things, into the one or more render targets that we provide to the OM, we only have one render target for now.
+The other two stages are not programmable but they are fairly easy to understand and configure:
 
-### Vertex Shader
+- the **Input Assembler** is responsible for processing the vertices in an eventual vertex buffer into the primitive topology of our choice, which in our case is a form of triangles, and sending this processed output to the Vertex Shader to process.
```
[the remark this answers is not in the document]
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md]: maybe simpler is a more accurate (since we will get into specifics later): `processed output to be processed again - but this time by our Vertex Shader.

this also highlights how we input to prepare for our own logic to do as we please with data

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 4 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:46]: I agree that usually is not emphasizing how when programming on the GPU, "the world is [still] your oyster", I still want to hint at the broad ways you can do things if we do this change though-

The vertices are **usually** read from a **Vertex Buffer** — usually since there are no
hard rules in programming, we can always be inventive to create or read data differently on the fly!
This all works since the vertex shader will be run however many times we tell 
it to run, which is specified in the first parameter of `ID3D11DeviceContext::Draw()` 
(more on this later), for instance if we call `Draw(3, 0)`, the vertex shader will 
run 3 times.

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 5 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:51]: I am confused at your comment- I will specify a bit more though since I think that helps with your comment?

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 4 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:46]: I can also reorder the full screen triangle example this way

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 6 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:56]: this can be put under the "more options to send vertex data exists" idea

just to give a generic idea on how things can be done - not how (since as you said - it is kinda out of scope)

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 9 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:103]: probably a bad time to explain all of this - the range of expected [-1.0, 1.0] is a fixed number - 
"if the `SV_Position` values are tried to be given a position outside the range of 
[-1.0, 1.0] they are clipped and we won't see them on screen. Due to this we need
to create our own math to transform any coordinates we have to a normalized [-1.0, 1.0]
This is going to be explored later in more depth."

due to this I think we should bring up that there is math we use later - since right now we are focusing on getting a triangle to draw, normalizing coordinates are  much better lumped up with the other explanations of  screen space, clip space, world space, etc. - explain it all once, not twice in 2 areas is my point

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 10 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:109]: I think structs are a better convention as well- plus if we explain it with an example of separately it may be very confusing later on with how we explain things using a frame of reference

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 11 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:114]: its important information, if you make a RW buffer in a pixel shader its important to know it will be a block, not a 'smooth' triangle for what is calculated

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 12 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:148]: 100% agree - this is a miscommunication

File: docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md (hunk 13 above)
Comment (16-Bit-Dog) [reply docs/1-introduction/1-1-getting-started/1-1-3-hello-triangle.md:166]: yeah... I don't see why that sentence is there... its not even vague - it is incomplete

**Review by deccer — APPROVED** [review]

**Review by Themperror — APPROVED** [review]
